### ISSUE TITLE
Execute EXISTS predicate as SEMI JOIN

### ISSUE THREAD

**Issue by kokosing** [prestodb/presto#6384]
Take a look at simple test on TPCH (with sf1). Below queries represent the same selection:
1. Semi join

```
presto:sf1> select count(*) from lineitem where orderkey IN (SELECT orderkey from lineitem);
```

Finished in 5 seconds with peak memory 44.13MB.
1. Current EXISTS implementation

```
select count(*) from lineitem o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);
```

Did not finish within 3 minutes and kept about 200MB peek memory during the most time of the test.

Plan for 1:

```
presto:sf1> explain select count(*) from lineitem where orderkey IN (SELECT orderkey from lineitem);
                                                              Query Plan                                                              
--------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [count:bigint]                                                                                                    
         _col0 := count                                                                                                               
     - Aggregate(FINAL) => [count:bigint]                                                                                             
             count := "count"("count_25")                                                                                             
         - RemoteExchange[GATHER] => count_25:bigint                                                                                  
             - Aggregate(PARTIAL) => [count_26:bigint]                                                                                
                     count_26 := "count"(*)                                                                                           
                 - Filter["semijoin_result"] => [orderkey:bigint, semijoin_result:boolean]                                            
                     - Project => [orderkey:bigint, semijoin_result:boolean]                                                          
                         - SemiJoin[orderkey = orderkey_1] => [orderkey:bigint, $hashvalue:bigint, semijoin_result:boolean]           
                             - Project => [orderkey:bigint, $hashvalue:bigint]                                                        
                                     $hashvalue := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))         
                                 - TableScan[tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey:bigint]                
                                         orderkey := tpch:orderkey                                                                    
                             - RemoteExchange[REPARTITION - REPLICATE NULLS] => orderkey_1:bigint, $hashvalue_27:bigint               
                                 - Project => [orderkey_1:bigint, $hashvalue_28:bigint]                                               
                                         $hashvalue_28 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey_1"), 0))
                                     - TableScan[tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey_1:bigint]          
                                             orderkey_1 := tpch:orderkey                                                              

(1 row)
```

Plan for 2: 

```
presto:sf1> explain select count(*) from lineitem o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);
                                                                                           Query Plan                                 
--------------------------------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [count_21:bigint]                                                                                                 
         _col0 := count_21                                                                                                            
     - Aggregate(FINAL) => [count_21:bigint]                                                                                          
             count_21 := "count"("count_26")                                                                                          
         - RemoteExchange[GATHER] => count_26:bigint                                                                                  
             - Aggregate(PARTIAL) => [count_27:bigint]                                                                                
                     count_27 := "count"(*)                                                                                           
                 - Project => []                                                                                                      
                     - Filter[("count" > BIGINT '0')] => [orderkey:bigint, unique:bigint, count:bigint]                               
                         - Project => [orderkey:bigint, unique:bigint, count:bigint]                                                  
                             - Aggregate[orderkey, unique] => [orderkey:bigint, unique:bigint, $hashvalue:bigint, count:bigint]       
                                     count := "count"("non_null")                                                                     
                                 - Project => [orderkey:bigint, unique:bigint, orderkey_0:bigint, non_null:boolean, $hashvalue:bigint]
                                         $hashvalue := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("order
                                     - LeftJoin[("orderkey_0" = "orderkey")] => [orderkey:bigint, unique:bigint, orderkey_0:bigint, no
                                         - RemoteExchange[REPARTITION] => orderkey:bigint, unique:bigint                              
                                             - AssignUniqueId => [orderkey:bigint, unique:bigint]                                     
                                                 - TableScan[tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey:bigint]
                                                         orderkey := tpch:orderkey                                                    
                                         - RemoteExchange[REPARTITION] => orderkey_0:bigint, non_null:boolean                         
                                             - Project => [non_null:boolean, orderkey_0:bigint]                                       
                                                     non_null := true                                                                 
                                                 - TableScan[tpch:tpch:lineitem:sf1.0, originalConstraint = true] => [orderkey_0:bigin
                                                         orderkey_0 := tpch:orderkey                                                  

(1 row)
```

The main difference is that IN predicate creates semi join, while EXISTS creates left outer join and aggregation (with high number of groups) on top of that. Semantically both queries means the same and in ideal world should (if possible) evaluate to the same optimal plan. So current plan is to implement execution of EXISTS as semi join. 

Above will require changes in semi join: 
- support multiple columns equality predicate
- support filter join filter function

Also might require https://github.com/prestodb/presto/issues/12154

**Comment by kokosing** [comment]
The work on this is inspired by the low performance and high resource usage of TPC-H q21.

**Comment by sopel39** [comment]
We could do this slightly differently.
Currently correlated subqueries with equi-conditions are rewritten to (more or less):
```
InnerJoin[equi_conditions](produces_only_probe_symbols)
 - probe
 - Aggregate(FINAL)[grouping_keys]()
   - RemoteExchage
     - Aggregate(PARTIAL)[grouping_keys]()
       - subquery
```
we could translate that into (after CBO reorder joins):
```
SemiJoin[equi_conditions]
  - probe
  - RemoteExchange
    - Aggregate(PARTIAL)[grouping_keys]()
      - subquery
```
This shouldn't introduce any perf regression

FYI: @mbasmanova

**Comment by mbasmanova** [comment]
@sopel39 @kokosing The plan for EXISTS query is different from the above. Is it still inefficient? BTW, do you know what's preventing optimizer from seeing that `EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey)` is always true and dropping that filter altogether?
```
presto:tiny> explain (type distributed) select count(*) from lineitem o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);
explain (type distributed) select count(*) from lineitem o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);
                                                                              Query Plan                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]                                                                                                                                                  
     Output layout: [count]                                                                                                                                           
     Output partitioning: SINGLE []                                                                                                                                   
     Execution Flow: UNGROUPED_EXECUTION                                                                                                                              
     - Output[_col0] => [count:bigint]                                                                                                                                
             _col0 := count                                                                                                                                           
         - Aggregate(FINAL) => [count:bigint]                                                                                                                         
                 count := "count"("count_25")                                                                                                                         
             - LocalExchange[SINGLE] () => count_25:bigint                                                                                                            
                 - RemoteSource[1] => [count_25:bigint]                                                                                                               
                                                                                                                                                                      
 Fragment 1 [tpch:orders:15000]                                                                                                                                       
     Output layout: [count_25]                                                                                                                                        
     Output partitioning: SINGLE []                                                                                                                                   
     Execution Flow: UNGROUPED_EXECUTION                                                                                                                              
     - Aggregate(PARTIAL) => [count_25:bigint]                                                                                                                        
             count_25 := "count"(*)                                                                                                                                   
         - InnerJoin[("orderkey" = "orderkey_0")][$hashvalue, $hashvalue_26] => []                                                                                    
                 Distribution: PARTITIONED                                                                                                                            
             - ScanProject[table = tpch:tpch:lineitem:sf0.01, originalConstraint = true] => [orderkey:bigint, $hashvalue:bigint]                                      
                     Cost: {rows: 60175 (587.65kB), cpu: 601750.00, memory: 0.00, network: 0.00}/{rows: 60175 (1.15MB), cpu: 1805250.00, memory: 0.00, network: 0.00} 
                     $hashvalue := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("orderkey"), 0))                                                         
                     orderkey := tpch:orderkey                                                                                                                        
             - LocalExchange[HASH][$hashvalue_26] ("orderkey_0") => orderkey_0:bigint, $hashvalue_26:bigint                                                           
                 - RemoteSource[2] => [orderkey_0:bigint, $hashvalue_27:bigint]                                                                                       
                                                                                                                                                                      
 Fragment 2 [tpch:orders:15000]                                                                                                                                       
     Output layout: [orderkey_0, $hashvalue_28]                                                                                                                       
     Output partitioning: tpch:orders:15000 [orderkey_0]                                                                                                              
     Execution Flow: UNGROUPED_EXECUTION                                                                                                                              
     - Project[] => [orderkey_0:bigint, $hashvalue_28:bigint]                                                                                                         
             Cost: {rows: 15000 (292.97kB), cpu: 1503500.00, memory: 150000.00, network: 0.00}                                                                        
             $hashvalue_28 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("orderkey_0"), 0))                                                            
         - Aggregate(STREAMING)[orderkey_0] => [orderkey_0:bigint]                                                                                                    
                 Cost: {rows: 15000 (146.48kB), cpu: 1203500.00, memory: 150000.00, network: 0.00}                                                                    
             - TableScan[tpch:tpch:lineitem:sf0.01, originalConstraint = true] => [orderkey_0:bigint]                                                                 
                     Cost: {rows: 60175 (587.65kB), cpu: 601750.00, memory: 0.00, network: 0.00}                                                                      
                     orderkey_0 := tpch:orderkey                                                                                                                      
```

**Comment by findepi** [comment]
> do you know what's preventing optimizer from seeing that `EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey)` is always true and dropping that filter altogether?

I understand this is in context where `o` stands for `lineitem`, i.e.
`select count(*) from lineitem o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);`
?

There is a problem with semantics -- if `lineitem` is mentioned twice in the query, it is definitely the same table. But, is it the same _contents_? What if a connector, on every read independently, returns 1000 random numbers? 
Of course, this example is rather contrived, but I think when we "de-duplicate" table scans, we should also add SPI way to opt-out from this. cc @martint

**Comment by atris** [comment]
I will work on this, thanks

**Comment by sopel39** [comment]
> There is a problem with semantics -- if lineitem is mentioned twice in the query, it is definitely the same table. But, is it the same contents? What if a connector, on every read independently, returns 1000 random numbers?

Also usually correlated subqueries are querying different table or the same table but with additional predicates. The example here trivial for demonstration purposes and not something what happens in practice.

> The plan for EXISTS query is different from the above. Is it still inefficient?

TPCH connector provides special properties for `lineitem.orderkey` column (`localProperties = ImmutableList.of(new SortingProperty<>(orderKeyColumn, SortOrder.ASC_NULLS_FIRST));`). I don't think this will happen in practice with Hive connector for example, so STREAMING aggregation won't be applied.

**Comment by kokosing** [comment]
> Of course, this example is rather contrived, but I think when we "de-duplicate" table scans, we should also add SPI way to opt-out from this.

@findepi As I understand "de-duplication" is out of scope of this ticket.

> TPCH connector provides special properties for lineitem.orderkey column (localProperties = ImmutableList.of(new SortingProperty<>(orderKeyColumn, SortOrder.ASC_NULLS_FIRST));). I don't think this will happen in practice with Hive connector for example, so STREAMING aggregation won't be applied.

@sopel39 You can have sorted tables in hive, but AFAIK Presto does not leverage that today.

**Comment by martint** [comment]
> There is a problem with semantics -- if lineitem is mentioned twice in the query, it is definitely the same table. But, is it the same contents? What if a connector, on every read independently, returns 1000 random numbers?

If a table is referenced twice in the same query, it should return the same contents (similar to WITH queries). Unfortunately, this can’t always be guaranteed by our connectors. As you suggest above, at some point we should allow connectors to provide additional metadata indicating tables are “non-deterministic” or something similar.the engine can then choose to materialize the table at some stage in the query instead of scanning it multiple times if it deems it cheaper to do that instead.

**Comment by findepi** [comment]
> As I understand "de-duplication" is out of scope of this ticket.

@kokosing yeah, it's not. The topic was brought up here in response to @mbasmanova's:

> BTW, do you know what's preventing optimizer from seeing that `[ select count(*) from lineitem o where ] EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey)` is always true and dropping that filter altogether?

**Comment by sopel39** [comment]
> BTW, do you know what's preventing optimizer from seeing that [ select count(*) from lineitem o where ] EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey) is always true and dropping that filter altogether?

I think the fact that such query just doesn't occur in practice, like uncorrelated exists

**Comment by mbasmanova** [comment]
@kokosing The query in the description is not practical. Is there another query that illustrates the problem and can be used to evaluate potential solutions?

**Comment by kokosing** [comment]
@mbasmanova I think I started with:
```
select count(*) from orders o where EXISTS (SELECT orderkey from lineitem WHERE orderkey = o.orderkey);
```

IIRC then I might I realize that `orders` was somohow to small, so I changed it to `lineitem`.  The thing that probe and build side scans the same table is irrelevant for this issue.

**Comment by atris** [comment]
@kokosing @findepi @sopel39 I am starting to work on this, and wanted to understand the approach that I need to take to implement the feature:

1) Support for multiple columns in Semijoin -- Is that already being done in #6385 ?

2) Per @sopel39 's comment above, the idea is to have correlated subqueries with equi-predicates as Semijoins and not Join + Agg. So the rule is not limited to just EXISTS clause, correct?

**Comment by findepi** [comment]
> Support for multiple columns in Semijoin -- Is that already being done in #6385 ?

i think @arhimondr is working on this.

**Comment by atris** [comment]
@findepi Should I pick this up after #6385 is done, then? Having multiple columns support in SemiJoin is a prerequisite for this.

**Comment by kokosing** [comment]
I think you don't need to wait, you can add a rewrite to semi join on single column. Once #6385 got done we could extend it.

**Comment by sopel39** [comment]
@atris 

> Per @sopel39 's comment above, the idea is to have correlated subqueries with equi-predicates as Semijoins and not Join + Agg. So the rule is not limited to just EXISTS clause, correct?

The idea is to rewrite `Join + Agg` into `SemiJoin + Partial aggregation` as in: https://github.com/prestodb/presto/issues/6384#issuecomment-398660235

**Comment by findepi** [comment]
> Execute EXISTS predicate as SEMI JOIN

Currently the EXISTS can, sometimes, be executed as an InnerJoin without aggregation and reordered.
Semi Join cannot (partially addressed by https://github.com/prestodb/presto/issues/12154)
